src/evaluate/evaluator.py

```diff
@@ -12,95 +12,125 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from abc import ABC, abstractmethod
 from numbers import Number
-from typing import Dict, Optional, Union
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union
 
 # Lint as: python3
-import numpy as np
-from datasets import Dataset, DatasetDict, IterableDataset, IterableDatasetDict, load_dataset, load_metric
+from datasets import Dataset, load_dataset, load_metric
 from datasets.metric import Metric
 from scipy.stats import bootstrap
 from transformers import Pipeline, pipeline
+from typing_extensions import Literal
 
 
-class Evaluator:
-    def __init__(
+class Evaluator(ABC):
+    def __init__(self, task, default_metric=None):
+        self.task = task
+        self.default_metric = default_metric
+
+    @abstractmethod
+    def _compute_predictions(self, pipe: Pipeline, inputs, **predictions_parameters: Dict):
+        raise NotImplementedError()
+
+    @staticmethod
+    def _compute_confidence_interval(
+        predictions,
+        references,
+        metric: Metric,
+        metric_keys: List[str],
+        confidence_level: float = 0.95,
+        n_resamples: int = 9999,
+    ) -> Dict[str, Any]:
+        bootstrap_dict = {}
+        for key in metric_keys:
+            bootstrap_dict[key] = bootstrap(
+                data=(predictions, references),
+                statistic=lambda predictions, references: metric.compute(
+                    predictions=predictions, references=references
+                )[key],
+                paired=True,
+                vectorized=False,
+                confidence_level=confidence_level,
+                n_resamples=n_resamples,
+            )
+        return bootstrap_dict
+
+    @abstractmethod
+    def compute(
         self,
-        pipe: Union[str, Pipeline],
-        data: Union[str, DatasetDict, Dataset, IterableDatasetDict, IterableDataset],
-        metric: Union[str, Metric],
-        metric_key: Optional[str] = None,
-        col_mapping: Optional[Dict[str, str]] = None,
-        label_mapping: Optional[Dict[str, Number]] = None,
-    ):
-        self.pipe = pipeline(pipe) if isinstance(pipe, str) else pipe
-        self.data = load_dataset(data) if isinstance(data, str) else data
-        self.metric = load_metric(metric) if isinstance(metric, str) else metric
-        self.metric_key = metric_key
-        self.label_mapping = label_mapping
-        if col_mapping is not None:
-            self.data = data.rename_columns(col_mapping)
-
-    def get_predictions(
-        self, baseline_pipe: Optional[Pipeline] = None, baseline_label_mapping: Optional[Dict[str, Number]] = None
+        model: str = None,
+        pipe: Union[Pipeline, Callable] = None,
```

Comment:
Having two arguments for the model/pipeline is not ideal I think: It makes it a bit less clear for a user what needs to be provided (a model, a pipe, or both?). What do you think about having it 
```suggestion
        model_or_pipeline: Union[Pipeline, Callable, str] = None,
```

Not that in the `transformers` ecosystem, especially around the pipelines users can usually pass a model instance or a string as model. So we should check if a user passed a model instance.